### DISCUSSION TITLE
Fallibility of starting a trillium server

### DISCUSSION THREAD

**Discussion by quixoticaxis** [trillium-rs/trillium#198] [Q&A] [answered]
Hello.

The question arose while I was trying to implement a toy book library utility that provides an optional web interface. The runner's `run` method does not return `Result<()>`, so the inability to bind to the address becomes a hard error. Is it intended or a work-in-progress API limitation?

**Comment by jbr** [comment, marked as answer]
One aspect of trillium's modularity is that components like the runtime adapters are intended to be modular as well. The current runtime adapters are primarily intended for ease of use for running servers where if they cannot bind to the specified host/port, the process should exit. It would not be difficult to either extend the current adapters to support fallible server starting, or build a new adapter that supports that special use case. A goal for trillium is to have a specific production use case in mind when making changes to an interface that impacts all users.  What would you want to happen if you couldn't bind to the specified host and port?

**Reply by quixoticaxis** [reply]
I personally would like to be able to handle it on my side, either by providing new host and port, or by deciding to disable the server's functionality and continue on without it.

**Reply by jbr** [reply]
Can you explain the use case a bit more? I'm trying to figure out how common this is, and whether this is the sort of thing that should be a custom / new runtime adapter, or whether it's worth changing the existing runtime adapters to support it. Trillium was designed primarily for servers that either start or fail to start (like a microservice deployed on cloud instances), but I can see the utility of supporting an escape hatch that's less usable but more powerful for more complex/advanced logic

**Reply by jbr** [reply]
I think the escape hatch probably would look like passing a TcpListener (or multiple TcpListeners, potentially) into config somehow, allowing you to perform whatever conditional logic *before* calling run.

**Reply by jbr** [reply]
In the situation where it's an optional server, why wouldn't you want the user to specify that? If you're looking to use a random but available port, you can use something like [portpicker](https://docs.rs/portpicker/latest/portpicker/)

**Reply by quixoticaxis** [reply]
My use case I've mentioned previously, although not production quality, is a service that enumerates archives with the books, and enriches the files with additional metadata (author, cover, etc.) from a specialized file and queries Internet in case the metadata is missing. Because I'm currently trying out Trillium, I wanted to host a statistics portal that would serve a simple HTML with progress reports. 

I believe any infrastructure service distributed as a single executable would do, I suppose. Off the top of my head I can imagine a few:
- A service that handles live back-ups of the disks and may run an optional simple HTTP server to serve statistics.
- A database that can optionally host it's own monitoring portal.

**Reply by quixoticaxis** [reply]
> In the situation where it's an optional server, why wouldn't you want the user to specify that? If you're looking to use a random but available port, you can use something like [portpicker](https://docs.rs/portpicker/latest/portpicker/)

Sure thing, it's unsafe though, because the port being free does not mean it would be free at the time of binding the server.
For my current case it's a non-issue, because I know the port is free. I was more interested in the API design and assumed that the behavior of crashing the process on binding failure is a bit too extreme in general case.

**Reply by quixoticaxis** [reply]
> In the situation where it's an optional server, why wouldn't you want the user to specify that? If you're looking to use a random but available port, you can use something like [portpicker](https://docs.rs/portpicker/latest/portpicker/)

I assume that asking the user for the port is fine if the end product is a technical utility. If it's a product targeted at not IT audience, I suppose, it is no longer viable.

**Reply by jbr** [reply]
> I believe any infrastructure service distributed as a single executable would do, I suppose. Off the top of my head I can imagine a few:

> A service that handles live back-ups of the disks and may run an optional simple HTTP server to serve statistics.
> A database that can optionally host it's own monitoring portal.

I would have expected these to accept command line configuration for the host and port (or to enable/disable the server). Alternatively, if you don't specify a port/host for trillium, it will pick up HOST and PORT from the environment.

I think a good escape hatch that wouldn't make people write `run(…).unwrap()` everywhere would be accepting an already-bound TcpListener instead of a host and port. I'll take a look at adding this to Config!

**Reply by jbr** [reply]
> I assume that asking the user for the port is fine if the end product is a technical utility. If it's a product targeted at not IT audience, I suppose, it is no longer viable.

How does the user know how to visit this url if they haven't specified the port themselves? Does the CLI open a browser? This was the sort of thing that I couldn't figure out when I built out tide's listener api (which _is_ fallible, but all of the examples just unwrap the call that starts the server)

**Reply by quixoticaxis** [reply]
> How does the user know how to visit this url if they haven't specified the port themselves? Does the CLI open a browser? This was the sort of thing that I couldn't figure out when I built out tide's listener api (which _is_ fallible, but all of the examples just unwrap the call that starts the server)

Yep, the utility can always feed the url to the OS, which would open the default browser.

**Reply by quixoticaxis** [reply]
> I think a good escape hatch that wouldn't make people write `run(…).unwrap()` everywhere would be accepting an already-bound TcpListener instead of a host and port. I'll take a look at adding this to Config!

That would be great, and would leave an option of a zero-cost update if `std::net::TcpListener` is ever extended with additional capabilities.
